File: cryptography-cipher/src/main/java/dev/retrotv/crypto/cipher/stream/Chacha20.java
package dev.retrotv.crypto.cipher.stream;

import dev.retrotv.crypto.cipher.param.Param;
import dev.retrotv.crypto.cipher.param.ParamWithIV;
import dev.retrotv.crypto.cipher.result.Result;
import lombok.NonNull;

import org.bouncycastle.crypto.engines.ChaChaEngine;
import org.bouncycastle.crypto.params.KeyParameter;
import org.bouncycastle.crypto.params.ParametersWithIV;

import java.io.InputStream;
import java.io.OutputStream;

/**
 * ChaCha20 스트림 암호화 클래스 입니다.
 */
public class Chacha20 extends StreamCipher {
    private static final String REQUIRED_MESSAGE = "ChaCha20 모드는 ParamsWithIV 객체를 요구합니다.";

    public Chacha20() {
        this.engine = new ChaChaEngine(20);
    }

    @Override
    public Result encrypt(@NonNull byte[] data, @NonNull Param params) {
        if (!(params instanceof ParamWithIV)) {
            throw new IllegalArgumentException(REQUIRED_MESSAGE);
        }

        ParamWithIV paramWithIV = (ParamWithIV) params;
        this.engine.init(true, new ParametersWithIV(new KeyParameter(paramWithIV.getKey()), paramWithIV.getIv()));
        byte[] encryptedData = new byte[data.length];
        this.engine.processBytes(data, 0, data.length, encryptedData, 0);
        
        return new Result(encryptedData);
    }

    @Override
    public void encrypt(@NonNull InputStream input, @NonNull OutputStream output, @NonNull Param params) {
        if (!(params instanceof ParamWithIV)) {
            throw new IllegalArgumentException(REQUIRED_MESSAGE);
        }

        ParamWithIV paramWithIV = (ParamWithIV) params;
        this.engine.init(true, new ParametersWithIV(new KeyParameter(paramWithIV.getKey()), paramWithIV.getIv()));
        this.streamEncrypt(input, output);
    }

    @Override
    public Result decrypt(@NonNull byte[] encryptedData, @NonNull Param params) {
        if (!(params instanceof ParamWithIV)) {
            throw new IllegalArgumentException(REQUIRED_MESSAGE);
        }

        ParamWithIV paramWithIV = (ParamWithIV) params;
        this.engine.init(false, new ParametersWithIV(new KeyParameter(paramWithIV.getKey()), paramWithIV.getIv()));
        byte[] originalData = new byte[encryptedData.length];
        this.engine.processBytes(encryptedData, 0, encryptedData.length, originalData, 0);
        
        return new Result(originalData);
    }

    @Override
    public void decrypt(@NonNull InputStream input, @NonNull OutputStream output, @NonNull Param params) {
        if (!(params instanceof ParamWithIV)) {
            throw new IllegalArgumentException(REQUIRED_MESSAGE);
        }
        
        ParamWithIV paramWithIV = (ParamWithIV) params;
        this.engine.init(false, new ParametersWithIV(new KeyParameter(paramWithIV.getKey()), paramWithIV.getIv()));
        this.streamDecrypt(input, output);
    }
}

File: cryptography-cipher/src/main/kotlin/dev/retrotv/crypto/cipher/stream/Chacha20.kt
package dev.retrotv.crypto.cipher.stream

import dev.retrotv.crypto.cipher.param.Param
import dev.retrotv.crypto.cipher.param.ParamWithIV
import dev.retrotv.crypto.cipher.result.Result
import org.bouncycastle.crypto.engines.ChaChaEngine
import org.bouncycastle.crypto.io.CipherInputStream
import org.bouncycastle.crypto.io.CipherOutputStream
import org.bouncycastle.crypto.params.KeyParameter
import org.bouncycastle.crypto.params.ParametersWithIV
import java.io.InputStream
import java.io.OutputStream

/**
 * ChaCha20 스트림 암호화 클래스 입니다.
 */
class Chacha20 : StreamCipher() {
    private val requiredMessage = "ChaCha20 모드는 ParamsWithIV 객체를 요구합니다."
    init {
        this.engine = ChaChaEngine(20)
    }

    override fun encrypt(data: ByteArray, params: Param): Result {
        require(params is ParamWithIV) { requiredMessage }

        this.engine.init(true, ParametersWithIV(KeyParameter(params.key), params.iv))

        val encryptedData = ByteArray(data.size)
        this.engine.processBytes(data, 0, data.size, encryptedData, 0)

        return Result(encryptedData)
    }

    override fun encrypt(input: InputStream, output: OutputStream, params: Param) {
        require(params is ParamWithIV) { requiredMessage }

        this.engine.init(true, ParametersWithIV(KeyParameter(params.key), params.iv))

        val cos = CipherOutputStream(output, this.engine)
        val buffer = ByteArray(1024)
        var i: Int = input.read(buffer)
        while (i != -1) {
            cos.write(buffer, 0, i)
            i = input.read(buffer)
        }

        cos.flush()
        cos.close()
    }

    override fun decrypt(encryptedData: ByteArray, params: Param): Result {
        require(params is ParamWithIV) { requiredMessage }

        this.engine.init(false, ParametersWithIV(KeyParameter(params.key), params.iv))

        val originalData = ByteArray(encryptedData.size)
        this.engine.processBytes(encryptedData, 0, encryptedData.size, originalData, 0)

        return Result(originalData)
    }

    override fun decrypt(input: InputStream, output: OutputStream, params: Param) {
        require(params is ParamWithIV) { requiredMessage }

        this.engine.init(false, ParametersWithIV(KeyParameter(params.key), params.iv))

        val cis = CipherInputStream(input, this.engine)
        val buffer = ByteArray(1024)
        var i: Int = cis.read(buffer)
        while (i != -1) {
            output.write(buffer, 0, i)
            i = cis.read(buffer)
        }

        cis.close()
    }
}